docs/RFCS/system_jobs.md

```diff
@@ -0,0 +1,424 @@
+`system.jobs` table in [#11722] as an example of using the new framework. His PR
+was rejected because the table's schema hadn't been fully considered. This RFC
+describes a revised `system.jobs` schema for us to thoroughly vet before we
+proceed with a final implementation.
+
+## The proposal
+
+Within this RFC, the term "job" or "long-running job" refers only to backups,
+restores, and schema changes. Other types of long-running queries, like slow
+`SELECT` statements, are explicitly out of scope. A "job-creating query," then,
+is any query of one of the following types:
+
+- `ALTER TABLE`, which creates a schema change job
+- `BACKUP`, which creates a backup job
+- `RESTORE`, which creates a restore job
+
+To track the progress of these jobs, the following table will be injected into
+the `system` database using the [cluster upgrade framework]:
+
+```sql
+CREATE TABLE system.jobs (
+    id               INT DEFAULT unique_rowid() PRIMARY KEY,
+    status           STRING NOT NULL,
+    percentCompleted DECIMAL(5, 2),
+    payload          BYTES,
+    INDEX status_idx (status)
+)
+```
+
+Each job is identified by a unique `id`, which is assigned when the job is
+created. Currently, this ID serves only to identify the job to the user, but
+future SQL commands to e.g. abort running jobs will need this ID to
+unambiguously specify the target job.
+
+The `status` column reprents a state machine with `pending`, `running`,
+`succeeded`, and `failed` states. Jobs are created in the `pending` state when
+the job-creating query is accepted, move to the `running` state once work on the
+job has actually begun, then move to a final state of `succeeded` or `failed`.
+Once a job has entered a final state, it becomes immutable. The `pending` state
+warrants additional explanation: it's used to track jobs that are enqueued but
+not currently performing work. Schema changes, for example will sit in the
+`pending` state until all prior schema change jobs have completed.
+
+The `percentCompleted` field, as expected, starts as `NULL` and is periodically
+updated from 0 to 100 while the job is `running`. Jobs in the `succeeded` state
+will always have a `percentCompleted` of 100, while jobs in the `failed` state
+may have any `percentCompleted` value. This value is stored as a decimal with a
+scale of two to allow tracking sub-percent progress on large jobs.
+
+We want to avoid future schema changes to `system.jobs` if at all possible.
+Every schema change requires a cluster upgrade, and every cluster upgrade
+introduces node startup time overhead, plus risk and complexity. To that end,
+fields that are not fundamental to the schema are stashed in the `payload`
+column, which stores a protobuf that can be evolved per the usual protobuf
+forwards-compatibility support. A potential structure of the `payload` follows.
+
+```protobuf
+message BackupJobPayload {
+    string to = 1;
+    repeated Descriptor targets = 2;
```

Comment:
Switched over to using IDs. Argument for not pulling out to `JobPayload`: it seems to me largely coincidental that every job is associated with one or more tables. I could see a job that's not table-related being tracked in this table (cluster-wide GC or something?). But you'd know better than I!